docs: add colab column (#583)
jina-ai/finetuner: docs/notebooks/text_to_image.md

---
jupyter:
  jupytext:
    text_representation:
      extension: .md
      format_name: markdown
      format_version: '1.3'
      jupytext_version: 1.14.1
  kernelspec:
    display_name: Python 3
    name: python3
---

<!-- #region id="3UCyCMPcvLGw" -->
# Text-to-Image Search via CLIP

<a href="https://colab.research.google.com/drive/1yKnmy2Qotrh3OhgwWRsMWPFwOSAecBxg?usp=sharing"><img alt="Open In Colab" src="https://colab.research.google.com/assets/colab-badge.svg"></a>

> Traditionally, searching images from text (text-image-retrieval) rely heavily on human annotations, this is commonly referred as *Text/Tag based Image Retrieval (TBIR)*.

Reviewer: ```suggestion
Traditionally, searching images from text (text-image-retrieval) relies heavily on human annotations, this is commonly referred to as *Text/Tag based Image Retrieval (TBIR)*.
```

> The [OpenAI CLIP](https://github.com/openai/CLIP) model maps the dense vector extracted from text and image into the same semantic space and produced a strong zero-shot model to mearesure the similarity between text and images.

Reviewer: ```suggestion
The [OpenAI CLIP](https://github.com/openai/CLIP) model maps the dense vectors extracted from text and image into the same semantic space and produces a strong zero-shot model to measure the similarity between text and images.
```

train_data.summary()
```

<!-- #region id="AE87a5Nvwd7q" -->
## Backbone model
Currently, we support several CLIP variations from [open-clip](https://github.com/mlfoundations/open_clip) for text to image retrieval tasks.

However, you can see all available models either in [choose backbone](https://finetuner.jina.ai/walkthrough/choose-backbone/) section or by calling `finetuner.describe_models()`.
<!-- #endregion -->

<!-- #region id="81fh900Bxgkn" -->
## Fine-tuning

Now that we have the training and evaluation datasets loaded as `DocumentArray`s and selected our model, we can start our fine-tuning run.
<!-- #endregion -->

```python id="UDcpfybOv1dh"
import finetuner

run = finetuner.fit(
    model='openai/clip-vit-base-patch32',
    train_data='fashion-train-data-clip',
    eval_data='fashion-eval-data-clip',
    epochs=5,
    learning_rate= 1e-7,
    loss='CLIPLoss',
    device='cuda',
)
```

<!-- #region id="QPDmFdubxzUE" -->
Let's understand what this piece of code does:

* We start with providing `model`, names of training and evaluation data.
* We also provide some hyper-parameters such as number of `epochs` and a `learning_rate`.

> * We use `CLIPLoss` to optimize CLIP model.

Reviewer: ```suggestion
* We use `CLIPLoss` to optimize the CLIP model.
```

<!-- #endregion -->

<!-- #region id="qKv3VcMKyG8d" -->
## Monitoring

> Now that we've created a run, let's see its status. You can monitor the run by checking the status - `run.status()`, the logs - `run.logs()` or `run.stream_logs()`. 

Reviewer: ```suggestion
Now that we've created a run, let's see its status. You can monitor the run by checking the status - `run.status()` - and the logs - `run.logs()` or `run.stream_logs()`. 
```

<!-- #endregion -->

```python id="JX45y-2fxs4L"
# note, the fine-tuning might takes 20~ minutes
for entry in run.stream_logs():
    print(entry)
```

<!-- #region id="xi49YlQsyXbi" -->
Since some runs might take up to several hours/days, it's important to know how to reconnect to Finetuner and retrieve your run.

```python
import finetuner

finetuner.notebook_login()
run = finetuner.get_run(run.name)
```

> You can continue monitoring the run by checking the status - `finetuner.run.Run.status()` or the logs - `finetuner.run.Run.logs()`.

Reviewer: ```suggestion
You can continue monitoring the run by checking the status - `finetuner.run.Run.status()` - or the logs - `finetuner.run.Run.logs()`.
```

```

```{admonition} Inference with ONNX
In case you set `to_onnx=True` when calling `finetuner.fit` function,
please use `model = finetuner.get_model(artifact, is_onnx=True)`
```
<!-- #endregion -->

<!-- #region id="LHyMm_M1zxdt" -->
## Advanced: WiSE-FT 

WiSE-FT, proposed by Mitchell et al. in [Robust fine-tuning of zero-shot models](https://arxiv.org/abs/2109.01903),
has been proven to be an effective way for fine-tuning models with a strong zero-shot capability,
such as CLIP.
As was introduced in the paper:

> Large pre-trained models such as CLIP or ALIGN offer consistent accuracy across a range of data distributions when performing zero-shot inference (i.e., without fine-tuning on a specific dataset). Although existing fine-tuning methods substantially improve accuracy on a given target distribution, they often reduce robustness to distribution shifts. We address this tension by introducing a simple and effective method for improving robustness while fine-tuning: ensembling the weights of the zero-shot and fine-tuned models (WiSE-FT).

Finetuner allows you to apply WiSE-FT easily,
all you need to do is use the `WiSEFTCallback`.

> Finetuner will trigger the callback when fine-tuning job finished and merge the weights between the pre-trained model and the fine-tuned model:

Reviewer: ```suggestion
Finetuner will trigger the callback when the fine-tuning job is finished and merge the weights between the pre-trained model and the fine-tuned model:
```
